File: src/org/stormgears/utils/dsio/SwitchControl.java
package org.stormgears.utils.dsio;

import edu.wpi.first.wpilibj.GenericHID;
import edu.wpi.first.wpilibj.buttons.JoystickButton;
import edu.wpi.first.wpilibj.command.Command;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import static org.apache.logging.log4j.util.Unbox.box;

public class SwitchControl extends JoystickButton {
	private static final Logger logger = LogManager.getLogger(SwitchControl.class);

	/**
	 * Create a joystick button for triggering commands.
	 *
	 * @param joystick     The GenericHID object that has the button (e.g. Joystick, KinectStick,
	 *                     etc)
	 * @param buttonNumber The button number (see {@link GenericHID#getRawButton(int) }
	 */
	public SwitchControl(@NotNull GenericHID joystick, int buttonNumber) {
		super(joystick, buttonNumber);

		super.whenReleased(new Command() {
			@Override
			protected boolean isFinished() {
				return true;
			}

			@Override
			protected void execute() {
				logger.info("Switch {} flipped off on {} ({})", box(buttonNumber), joystick.getName(), box(joystick.getPort()));
			}
		});

		super.whenPressed(new Command() {
			@Override
			protected boolean isFinished() {
				return true;
			}

			@Override
			protected void execute() {
				logger.info("Switch {} flipped on on {} ({})", box(buttonNumber), joystick.getName(), box(joystick.getPort()));
			}
		});
	}

	public void whenFlipped(@NotNull FlipListener listener) {
		super.whenReleased(new Command() {
			@Override
			protected boolean isFinished() {
				return true;
			}

			@Override
			protected void execute() {
				listener.flipped(false);
			}
		});

		super.whenPressed(new Command() {
			@Override
			protected boolean isFinished() {
				return true;
			}

			@Override
			protected void execute() {
				listener.flipped(true);
			}
		});
	}

	@FunctionalInterface
	public static interface FlipListener {
		void flipped(boolean on);
	}
}

File: src/main/java/org/stormgears/utils/dsio/SwitchControl.kt
package org.stormgears.utils.dsio

import edu.wpi.first.wpilibj.GenericHID
import edu.wpi.first.wpilibj.buttons.JoystickButton
import edu.wpi.first.wpilibj.command.Command
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.util.Unbox.box
import org.stormgears.utils.LambdaCommand

open class SwitchControl
/**
 * Create a joystick button for triggering commands.
 *
 * @param joystick     The GenericHID object that has the button (e.g. Joystick, KinectStick,
 * etc)
 * @param buttonNumber The button number (see [GenericHID.getRawButton]
 */
(private val joystick: GenericHID, private val buttonNumber: Int) : JoystickButton(joystick, buttonNumber), ISwitch {

	protected constructor(joystick: GenericHID, buttonNumber: Int, log: Boolean) : this(joystick, buttonNumber) {
		if (!log) {
			this.whenFlipped { isOn -> logger.trace("Switch {} flipped {} - {} ({})", box(buttonNumber), if (isOn) "on" else "off", joystick.name, box(joystick.port)) }
		}
	}

	override fun whenFlipped(listener: ISwitch.FlipListener) {
		super.whenReleased(object : Command() {
			override fun isFinished(): Boolean {
				return true
			}

			override fun execute() {
				listener.flipped(false)
			}
		})

		super.whenPressed(object : Command() {
			override fun isFinished(): Boolean {
				return true
			}

			override fun execute() {
				listener.flipped(true)
			}
		})
	}

	override fun whenFlipped(listener: (Boolean) -> Unit) {
		super.whenReleased(LambdaCommand { listener(false) })
		super.whenPressed(LambdaCommand { listener(true) })
	}

	companion object {
		private val logger = LogManager.getLogger(SwitchControl::class.java)
	}
}
